### PR TITLE
fix: enforce unique constraint on line item adjustment model and update service

### PR DESCRIPTION
**Pull request by zakariaelas**
### What

- enforces a composite unique constraints on `(discount_id, item_id)` where `discount_id is not null`
- extend the line item service to include a `generateAdjustments` method (see #1243 for the usage)

### CODE REVIEW

**Comment by olivermrbl** [comment]
Looks like this PR includes a lot of changes unrelated to what it solves. Could I get you to have a look at it @zakariaelas ?

**Comment by zakariaelas** [comment]
@olivermrbl fixed 😄 !

File: packages/medusa/src/models/line-item-adjustment.ts [hunk 1]
```diff
@@ -7,13 +7,15 @@ import {
...
   OneToOne,
+  Unique
 } from "typeorm"
 import { ulid } from "ulid"
 import { DbAwareColumn } from "../utils/db-aware-column"
 import { Discount } from "./discount"
 import { LineItem } from "./line-item"
 
 @Entity()
+@Index(["discount_id", "item_id"], { unique: true, where: `"discount_id" IS NULL` })
```
Comment (olivermrbl) [inline packages/medusa/src/models/line-item-adjustment.ts:18]: Should this not be `where: "discount_id" IS NOT NULL`?

Comment (olivermrbl) [reply packages/medusa/src/models/line-item-adjustment.ts:18]: nit: For other unique constraints, we've used the `@Unique` decorator like so:
```
@Unique("uniqitemdisc", ["discount_ite", "item_id"])
```

Comment (olivermrbl) [reply packages/medusa/src/models/line-item-adjustment.ts:18]: Would be great to have a test that validates the uniqueness constraint.

Comment (zakariaelas) [reply packages/medusa/src/models/line-item-adjustment.ts:18]: > Should this not be `where: "discount_id" IS NOT NULL`?

Yes definitely! Sorry, I had it on the `feat/line-item-adjustments-draft-orders` PR [here](https://github.com/medusajs/medusa/blob/be46e5cb14b88e23333cfa3c482e6845ac324d7f/packages/medusa/src/models/line-item-adjustment.ts#L18). Will fix in this branch too!

> nit: For other unique constraints, we've used the `@Unique` decorator like so:
> 
> ```
> @Unique("uniqitemdisc", ["discount_ite", "item_id"])
> ```

Will it enforce the constraint only if the `discount_id` is not null?  I tried using that at first, but couldn't find a way to tell typeorm to only enforce the unique constraint when `discount_id` is not null. Did you achieve something similar in the past using `@Unique()` ?



> Would be great to have a test that validates the uniqueness constraint.

Very true! Will add a few 👍

**Review by srindom — COMMENTED** [review]
Please fix failing tests

**Review by srindom — APPROVED** [review]
LGTM!

**Pull request merged**